### ISSUE TITLE
[ENH] Add preprocessing to `tslib` datamodule

### ISSUE THREAD

**Issue by phoeenniixx** [jdb78/pytorch-forecasting#2330]
Currently, we dont have preprocessing support in `tslib` datamodule. We should add that!
We could add the scalers and normalizers in a similar way as in #2302, but before that, I think we should 
look at  [thuml's](https://github.com/thuml/Time-Series-Library) implementation of pre-processing steps. We should add the implementation inspired from that package (as `tslib` datamodule is inspired from that package).

### How should we do this?
- First we should look at the [Time-Series-Library](https://github.com/thuml/Time-Series-Library), and then write down the how the preprocessing steps (like scaler and normalizer use) is implemented there
- Try to replicate that in `tslib` data module while staying coherent with the v2 design.

**Comment by Faakhir30** [comment]
**TODO:** refactor preprocessing of encoder-decoder data module and `tslib` to base where possible once this issue has a clear implementation.

Ref https://github.com/sktime/pytorch-forecasting/pull/2313

**Comment by echo-xiao** [comment]
hi @fkiraly please assign this to me.

**Comment by echo-xiao** [comment]
hi, @phoeenniixx , just wanna clarify my understanding towards thuml's preprocessing. 

thuml's full data pipeline: ingestion → temporal split → scaling (fit on train stats) → time features → windowing → optional augmentation/masking → collate → (then in-model: per-window norm / patching / decomposition, and eval-side:
inverse_transform etc.)

for the two current tslib models (DLinear, TimeXer), we need split + scaling + windowing + collate. Windowing and collate already exist; this PR adds scaling. the remaining gap is a within-series temporal split option. so, is adding the within-series temporal split enough for now? or should we also prepare for future models by adding time-feature encoding and an augumentation mechanism?


| Step | Status / plan |
|---|---|
| Temporal split | Currently random across series. i will add a thuml-style within-series temporal split option  |
| Train-statistics normalization | Done in this PR, mirroring `EncoderDecoderDataModule` (#2302) |
| Time-feature encoding | The pathway already exists: add calendar columns (hour, day-of-week, ...) to the DataFrame and mark them known → they flow into `future_cont` automatically. Auto-generation is missing and blocked on D1 datetime support, should users do it in the DataFrame themselves? |
| Window sampling | Already exists |
| Augmentation / on-the-fly masking | does augumentation or on-the-fly masking belong to preprocessing? should we implement augumentation as a train_transform hook? |
| Collate | Already exists; fixed-length windows just stack. The dict handling needed once `target_scale` lands belongs to that follow-up (sibling of #2359) |

**Comment by phoeenniixx** [comment]
> Currently random across series. i will add a thuml-style within-series temporal split option

Can you please look at #2087 - does it already cover your requirement? Maybe we can implement multiple split strategies in a new folder (like `splits`?)

> Auto-generation is missing and blocked on D1 datetime support, should users do it in the DataFrame themselves?

Can you please elaborate what do you mean by auto-generation?

> does augumentation or on-the-fly masking belong to preprocessing? should we implement augumentation as a train_transform hook?

I think we should add this! Maybe the implementation can live in `preprocessing` folder (if needed) or directly inside the DM - we can decide on it. What are your reservations about this?
Also tagging @fkiraly @agobbifbk for their thoughts!

**Comment by phoeenniixx** [comment]
I think we should not couple these feats into one PR as that will make it hard to review, we should move step by step - implementing just basic preprocessing and then adding other things on the top of it

One question here though - According to you, should we support all the scalers and normalizers we support in `EncoderDEcoderDataModule` in `tslib` one, or should we have some other custom implementations? idts `thuml`'s implementation uses `ptf` normalizers - Does it use `sklearn` ones or does it implements its own normalizers?

**Comment by echo-xiao** [comment]
1. time-feature encoding: 
- in thuml's utils/timefeatures.py, you pass a frequency string and it picks the calendar features for you (h -> HourOfDay) and scaled to [-0.5, 0.5], users don't need to write these
- our v2 version, users adds the columns and marks them known, and they flow into future_cnt, but the deriving part is missing. i think it's blocked at D1. TimeSeries requires time to be an integer column increamenting by 1, so there's no timestamp to derive from. so i suggest we don't do it now, we can open an issue on D1 later. 

2. augmentation. i do think we should have it. my only hesitation was wheather it belongs to the DM's preprocessing stage, not t as a separate step. thuml does it inside the Dataset. i think it should be a train_transform hook, train-split only and off by default. we can open a new issue for it later.

**Comment by echo-xiao** [comment]
#2087 cover it. thuml's implementation uses sklearn StandardScaler to fit on the train segment, the Normalizer writes itself is per-sample and calssfication only. since ScalerAdapter alreadysupports sklearn scalers, i think we should use ptf intead of customizing one. here is my future plan:

| Step | Status / plan |
| --- | --- |
| Temporal split | covered by #2087. i'll open a new PR to align the scaler fit boundary with it |
| Train-statistics normalization | this PR #2368  |
| Time-feature encoding | new D1 issue in the future |
| Window sampling | already exists |
| Augmentation / masking | a new issue |
| Collate | already exists |

**Comment by agobbifbk** [comment]
Thx @echo-xiao. All the thuml's have a huge hardcoded part related to temporal features. As you mention we can eventually add those in D1 in the future. I don't understand this, can you explain 

> our v2 version, users adds the columns and marks them known, and they flow into future_cnt, but the deriving part is missing. i think it's blocked at D1. TimeSeries requires time to be an integer column increamenting by 1, so there's no timestamp to derive from. so i suggest we don't do it now, we can open an issue on D1 later.

**Comment by phoeenniixx** [comment]
> * our v2 version, users adds the columns and marks them known, and they flow into future_cnt, but the deriving part is missing. i think it's blocked at D1. TimeSeries requires time to be an integer column increamenting by 1, so there's no timestamp to derive from. so i suggest we don't do it now, we can open an issue on D1 later.

Yes the auto-derivation of cols is missing which can happen by looking at the datatypes. We need to add that! I agree with you we should skip it for now.

For now, I would suggest just adding the support for scalers and normalizers, rest of the things can be added later as the package evolves

**Comment by phoeenniixx** [comment]
> [#2087](https://github.com/sktime/pytorch-forecasting/pull/2087) cover it. thuml's implementation uses sklearn StandardScaler to fit on the train segment, the Normalizer writes itself is per-sample and calssfication only. since ScalerAdapter alreadysupports sklearn scalers, i think we should use ptf intead of customizing one. here is my future plan:
> 
> Step	Status / plan
> Temporal split	covered by [#2087](https://github.com/sktime/pytorch-forecasting/pull/2087). i'll open a new PR to align the scaler fit boundary with it
> Train-statistics normalization	this PR [#2368](https://github.com/sktime/pytorch-forecasting/pull/2368)
> Time-feature encoding	new D1 issue in the future
> Window sampling	already exists
> Augmentation / masking	a new issue
> Collate	already exists

Looks good thanks!



> 2\. augmentation. i do think we should have it. my only hesitation was wheather it belongs to the DM's preprocessing stage

I think it belongs to DM, We may not want to add the augmentation as somthing "hardcoded" to `tslib` dm.
From Design And API standpoint, I think we should follow the strategy pattern here. These different techniques should be separate classes (like the `ptf` normalizers are seaprarte classes which reside in `data/encoders`). This will make it extensible. we can add `augmentation` or similar param to the dm, which will accept these classes and by default there will be `None`. It should act in a similar way as it happens for scalers and normalizers. I think it is possible right?

```python
dm = tslibdm(
...
augmentation = Jitter(),
)
```

Somthing like this.. But we should think of how this actually looks like - rn this is just a very crude design i can think of. We should look at the technicalities and think if this is possible. Here classes may take any input or something as well if needed.

But for now, i will keep augmentation in a separate PR (and issue), we should move step by step here.

Would you like to open an issue for augmentation where we can discuss the complete design for this?

**Comment by echo-xiao** [comment]
> All the thuml's have a huge hardcoded part related to temporal features. As you mention we can eventually add those in D1 in the future. I don't understand this, can you explain

sorry about the slow reply. i went back and checked this.

v2 won't derive calendar features for you. but you can add hour, day_of_week etc. as columns yourself and mark them known. and they flow through to future_cont on their own. so what is missing is the automation, not the capability. in thuml you pass freq='h' and utils/timefeature.py picks the right features and scales them to [-0.5, 0.5], in ptf you compute them yourself.

the actual blocker is in D2:
- v2's TimeSeries does accept a datatime column, v1 had assert dtype.kind=='i', v2 dropped it but the docstring still says integer. the timestamps are sitting right there in _preprocess_data. 
- they can't reach the batch because torch has no datetime dtype. 
- so they'd have to be turned into number first, and nothing does that. 

so, auto-derivation needs one conversion step in D2. D1 just needs the contract spelled out (does it supprot datetime time columns or not)

**Comment by echo-xiao** [comment]
hi, @phoeenniixx status update on this:

1. scalers are done in #2368. they are now fit on the training split only. target_normalizer is off by default until target_scale lands in #2359. 
2. augumentation: opened #2417 with a design proposal
